definition_validator/validator.cc

```diff
@@ -112,15 +112,34 @@ void matchPositional(const core::Context ctx, absl::InlinedVector<reference_wrap
     }
 }
 
-// Ensure that two argument lists are compatible in shape and type
+// Ensure that two argument lists are compatible in shape and type and that method visibility is compatible
 void validateCompatibleOverride(const core::Context ctx, core::MethodRef superMethod, core::MethodRef method) {
-    if (method.data(ctx)->flags.isOverloaded) {
+    auto methodData = method.data(ctx);
+
+    if (methodData->flags.isOverloaded) {
         // Don't try to check overloaded methods; It's not immediately clear how
         // to match overloads against their superclass definitions. Since we
         // Only permit overloading in the stdlib for now, this is no great loss.
         return;
     }
 
+    auto superMethodData = superMethod.data(ctx);
+
+    auto isPrivate = methodData->flags.isPrivate;
+    auto isProtected = methodData->flags.isProtected;
+    auto superIsProtected = superMethodData->flags.isProtected;
+    auto superIsPublic = superMethodData->isMethodPublic();
+
+    if ((isPrivate && (superIsProtected || superIsPublic)) || (isProtected && superIsPublic)) {
+        if (auto e = ctx.state.beginError(methodData->loc(), core::errors::Resolver::BadMethodOverride)) {
+            e.setHeader("Override of `{}` must be {} but `{}` is {}. Alternatively, `{}` can be made {}",
+                        superMethod.show(ctx), superIsProtected ? "protected or public" : "public", method.show(ctx),
+                        isPrivate ? "private" : "protected", superMethod.show(ctx),
+                        isPrivate ? "private" : "protected");
```

Comment:
I think the "Alternatively, ..." bit here is suboptimal for two reasons:

1. It makes an already long error message longer by repeating `superMethod`; and
2. There's not really a first course of action that it's providing an alternative for.  I guess the implicit course of action is that `method` should have its visibility flipped, but I don't think that's super clear.

I think if we *did* want to have the suggestion about how to correct the problem, we'd relegate it to an `e.addErrorLine` call.  But since there's two courses of action and we're not sure which one to choose, let's leave it out for now.

(I don't know if the suggestion can be applied as-is due to formatting lints.)

```suggestion
            e.setHeader("Override of `{}` must be {} but `{}` is {}.",
                        superMethod.show(ctx), superIsProtected ? "protected or public" : "public", method.show(ctx),
                        isPrivate ? "private" : "protected");
```

Reply:
> But since there's two courses of action and we're not sure which one to choose, let's leave it out for now.

It seems like the common case of `public` in parent (because it's an interface method) and `private` in child should be almost always be resolved by removing `private` from the child, not by adding `private` to the parent.

Defaulting the suggestion to removing `private` also has the nice property that it's unlikely to cause cascading type errors (meanwhile, asking the user to add `private` to the parent class might not work if the parent method is already being called publicly).

(This is basically what froydnj's suggestion implements; I just wanted to chime in my two cents.)

Reply:
Thanks for the feedback. I am completely happy to drop the second part of the message. I felt like it was a little verbose too but was trying to match the requirements/recommendation in the [original issue](https://github.com/sorbet/sorbet/issues/5522#issue-1185194258). I'll take another pass at this tomorrow night or maybe tonight if I get a few minutes.